community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java

```diff
@@ -294,52 +294,56 @@ public void apply( TransactionToApply batch, TransactionApplicationMode mode ) t
         }
     }
 
     /**
-     * Creates a {@link CommandApplierFacade} that is to be used for all transactions in a batch.
-     * Before starting to apply contents of each transaction a call to {@link CommandHandler#begin(TransactionToApply)}
-     * is issued, otherwise the appliers all just gets visit-calls with contents from all transactions.
+     * Creates a {@link BatchTransactionApplier.BatchTransactionApplierFacade} that is to be used for all transactions
+     * in a batch. Each transaction is handled by a {@link TransactionApplier.TransactionApplierFacade} which wraps the
+     * individual {@link TransactionApplier}s returned by the wrapped {@link BatchTransactionApplier}s.
      *
-     * After all transactions have been applied the appliers are closed (apply() then close()).
+     * After all transactions have been applied the appliers are closed.
      */
-    private CommandApplierFacade applier( TransactionApplicationMode mode )
+    private BatchTransactionApplier.BatchTransactionApplierFacade applier( TransactionApplicationMode mode )
     {
+        ArrayList<BatchTransactionApplier> appliers = new ArrayList<>();
         // Graph store application. The order of the decorated store appliers is irrelevant
-        CommandHandler storeApplier = new NeoStoreTransactionApplier( neoStores, cacheAccess, lockService );
+        appliers.add( new NeoStoreBatchTransactionApplier( neoStores, cacheAccess, lockService ) );
         if ( mode.needsHighIdTracking() )
         {
-            storeApplier = new HighIdTransactionApplier( storeApplier, neoStores );
+            appliers.add( new HighIdBatchTransactionApplier( neoStores ) );
         }
         if ( mode.needsCacheInvalidationOnUpdates() )
         {
-            storeApplier = new CacheInvalidationTransactionApplier( storeApplier, neoStores, cacheAccess );
+            appliers.add( new CacheInvalidationBatchTransactionApplier( neoStores, cacheAccess ) );
         }
 
         // Schema index application
-        CommandHandler indexApplier = new IndexTransactionApplier( indexingService, labelScanStoreSync );
+        appliers.add( new IndexBatchTransactionApplier( indexingService, labelScanStoreSync ) );
 
         // Legacy index application
-        CommandHandler legacyIndexApplier = new LegacyIndexApplier( indexConfigStore,
-                legacyIndexApplierLookup, legacyIndexTransactionOrdering, mode );
+        appliers.add(
+                new LegacyBatchIndexApplier( indexConfigStore, legacyIndexApplierLookup, legacyIndexTransactionOrdering,
+                        mode ) );
 
         // Counts store application
-        CommandHandler countsStoreApplier = new CountsStoreApplier( neoStores.getCounts(), mode );
+        appliers.add( new CountsStoreBatchTransactionApplier( neoStores.getCounts(), mode ) );
 
         // Perform the application
-        return new CommandApplierFacade( storeApplier, indexApplier, legacyIndexApplier, countsStoreApplier );
```

Comment:
Why was this changed to an ArrayList? Is there any benefit to it? If you ask me it's simply more garbage.